### PR TITLE
Proposal: Poetry setup

### PR DESCRIPTION
**Pull request by mooster531**
We could use [poetry](https://python-poetry.org/) to ship project configuration, which would make it easier to get up and running with a development setup after cloning the repo, as well as managing external dependencies inside the repository.

The following is a very quick crash course to Poetry, for reference.

# Poetry Quick Start Guide
A crash course in Poetry for experienced Python developers

## Installation
```bash
# macOS/Linux
curl -sSL https://install.python-poetry.org | python3 -

# Windows PowerShell
(Invoke-WebRequest -Uri https://install.python-poetry.org -UseBasicParsing).Content | py -

# Verify installation
poetry --version
```

## Key Concepts
Poetry manages project dependencies, virtual environments, and packaging in a single tool. It uses `pyproject.toml` for configuration and `poetry.lock` for dependency locking.

### Project Setup
```bash
# Start a new project
poetry new my-project

# Initialize poetry in existing project
poetry init

# Activate virtual environment
poetry shell
```

### Dependency Management
```bash
# Add dependencies
poetry add requests           # Add latest version
poetry add requests@^2.25.1   # Add specific version
poetry add --dev pytest      # Add dev dependency

# Remove dependencies
poetry remove requests

# Install from poetry.lock
poetry install

# Update dependencies
poetry update                # Update all
poetry update requests       # Update specific package
```

### Running Scripts
```bash
# Run a command in the virtual environment
poetry run python script.py
poetry run pytest

# Execute configured scripts from pyproject.toml
poetry run my-script
```

## pyproject.toml Example
```toml
[tool.poetry]
name = "my-project"
version = "0.1.0"
description = ""
authors = ["Your Name <you@example.com>"]

[tool.poetry.dependencies]
python = "^3.8"
requests = "^2.28.0"

[tool.poetry.dev-dependencies]
pytest = "^7.1.0"

[tool.poetry.scripts]
my-script = "my_project.cli:main"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

### Environment Management
```bash
# Configure virtualenv location
poetry config virtualenvs.in-project true  # Create venv in project directory

# Show path to virtual environment
poetry env info

# List all virtual environments
poetry env list

# Remove virtual environments
poetry env remove python3.8
```

### Dependency Resolution
```bash
# Show dependency tree
poetry show --tree

# Export requirements.txt
poetry export -f requirements.txt --output requirements.txt

# Export with dev dependencies
poetry export -f requirements.txt --with dev --output dev-requirements.txt
```

### Multiple Environment Dependencies
```toml
[tool.poetry.dependencies]
python = "^3.8"

[tool.poetry.group.dev.dependencies]
pytest = "^7.1.0"
black = "^22.3.0"

[tool.poetry.group.docs.dependencies]
sphinx = "^4.5.0"
```

## Best Practices
1. Always commit both `pyproject.toml` and `poetry.lock`
2. Use `poetry add` instead of manually editing `pyproject.toml`
3. Use groups to organize dependencies by environment
4. Specify Python version constraints appropriately
5. Use `poetry.lock` for reproducible builds

## Common Issues & Solutions
- If Poetry is slow: Use `poetry config installer.max-workers 10`
- Virtual environment conflicts: Delete `poetry.lock` and `.venv`, then run `poetry install`
- Path issues: Ensure Poetry's bin directory is in your PATH
- Dependencies not found: Try `poetry cache clear . --all`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,47 +1,89 @@
...
+]
+
+[tool.poetry.dependencies]
+python = "^3.13"
+dune-client = ">=1.7.7"
+pandas = "*"
+sqlalchemy = "*"
+python-dotenv = "*"
+psycopg2-binary = "*"
+tomli = "*"
```
Comment (bh2smith) [inline pyproject.toml]: I think we are missing pyyaml here:

https://github.com/bh2smith/dune-sync/blob/main/requirements/prod.txt


Shouldn't we also delete our requirements directory as well?

File: pyproject.toml [hunk 2]
```diff
@@ -1,47 +1,89 @@
+[project]
+name = "dune-sync"
+dynamic = ['version', 'dependencies']
+
+[tool.poetry]
+name = "dune-sync"
+version = "0.2.0"
```
Comment (bh2smith) [inline pyproject.toml]: I don't really want to have to update this manually. Can it be set to 0.0.0 or left out?

**Review by bh2smith — APPROVED** [review]
This seems cool and I would be happy to give it a try. However I think there seems to be some stuff that we should update accordingly. 

- makefile
- `rm -rf requirements/`
- Maybe the readme.

File: pyproject.toml (hunk 2 above)
Comment (mooster531) [reply pyproject.toml]: My bad, this is superfluous. The version is declared as a `dynamic` property and that's enough to make the tooling infer it from git:
```toml
[project]
name = "dune-sync"
dynamic = ['version', 'dependencies']
```

File: pyproject.toml (hunk 1 above)
Comment (mooster531) [reply pyproject.toml]: The contents of the `requirements` directory are still being used by the Dockerfile. I will adapt that to use poetry as well so we can remove it.

Sorry about the missing pyyaml dependency, I added it.

Comment (mooster531) [reply pyproject.toml]: Done in [817b0a2](https://github.com/bh2smith/dune-sync/pull/88/commits/817b0a28269dd238ae207ae1470fe2aa22fb0859)

**Comment by mooster531** [comment]
We now clear CI jobs too. Let me know if anything else stands out or needs changing. 
Wasn't sure if @fleupold would be okay with his email listed in the [pyproject.toml](https://github.com/bh2smith/dune-sync/pull/88/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R12) so I masked it. Please let me know if I should put some other (real) email there :)

**Comment by bh2smith** [comment]
I tried this out and everything looks pretty good here. We can merge this whenever.

**Pull request merged**